lib/src/components/NxCombobox/NxCombobox.tsx

```diff
@@ -0,0 +1,344 @@
+        long,
+        short,
+        disabled,
+        emptyMessage,
+        autoComplete,
+        inputProps,
+        id,
+        'aria-required': ariaRequired,
+        'aria-describedby': ariaDescribedBy,
+        ...attrs
+      } = props,
+
+      isEmpty = !matches.length,
+
+      dropdownRef = useRef<HTMLDivElement>(null),
+      inputRef = useRef<HTMLDivElement>(null),
+      elFocusedOnMostRecentRender = useRef<Element | null>(null),
+
+      [focusableBtnIndex, setFocusableBtnIndex] = useState<number | null>(null),
+      [elToFocusId, setElToFocusId] = useState(''),
+      [prevSearchText, setPrevSearchText] = useState(''),
+      [showDropdown, setShowDropdown] = useState(false),
+
+      inputId = useUniqueId('nx-combobox-input', id),
+      dropdownId = useUniqueId('nx-combobox-dropdown'),
+      dropdownRole = error || loading || isEmpty ? 'alert' : 'listbox',
+      dropdownBtnId = useUniqueId('nx-dropdown-button'),
+
+      inputClassName = classnames('nx-combobox__input', { 'nx-text-input--long': long,
+        'nx-text-input--short': short }),
+      className = classnames('nx-combobox', classNameProp, {
+        'nx-combobox--dropdown-showable': showDropdown
+      }),
+      dropdownClassName = classnames('nx-combobox__menu', {
+        'nx-combobox__menu--error': !!error
+      });
+
+  // There is a requirement that when there is an error querying the data, if the user navigates away from
+  // the component and then comes back to it the search should be retried automatically
+  function handleComponentFocus(evt: FocusEvent<HTMLDivElement>) {
+    // check if this is focus coming into the component from somewhere else on the page, not just moving between
+    // children of this component and not from focus coming into the browser from some other window
+    const comingFromOutsidePage = evt.relatedTarget === null,
+        comingFromChildNode = evt.relatedTarget instanceof Node && evt.currentTarget.contains(evt.relatedTarget);
+
+    if (!(comingFromOutsidePage || comingFromChildNode) && error) {
+      doSearch(searchText);
+    }
+    else if (error) {
+      setShowDropdown(true);
+    }
+  }
+
+  // In the blur event, we use the relatedTarget to check if the focused element is a child of the parent
+  // to show/hide dropdown, however, clicking the button in Safari does not focus it, which means the
+  // relatedTarget will be null, this is to stop the mouse down event being fired for the element that
+  // is listening to the blur event, in this case, the show/hide dropdown is handled by the click event
+  function handleMouseDown(evt: MouseEvent) {evt.preventDefault(); }
+
+  function handleComponentBlur(evt: FocusEvent<HTMLDivElement>) {
+    setElToFocusId('');
+    setFocusableBtnIndex(null);
+
+    // Check if the new focused element is a child of the parent, if not, then close the dropdown menu
+    if (!(evt.relatedTarget instanceof Node && evt.currentTarget.contains(evt.relatedTarget))) {
```

Comment:
In what circumstance is this true? If the component only has a single focusable element, then wouldn't any blur be a blur of that one element and only ever a focus of something outside the component?

Reply:
It was when clicking the item within the dropdown, the focus then moved to the item, the `showDropdown` should be set to false within `onClick` function instead.
However, I later created another function `function handleMouseDown(evt: MouseEvent) { evt.preventDefault(); }` to handle the focus issue in the Safari, which will stop the blur being fired when clicking the item, in that case, I think the if statement here could be removed. What do you think?

Reply:
> It was when clicking the item within the dropdown, the focus then moved to the item, the showDropdown should be set to false within onClick function instead.
However, I later created another function function handleMouseDown(evt: MouseEvent) { evt.preventDefault(); } to handle the focus issue in the Safari, which will stop the blur being fired when clicking the item, in that case, I think the if statement here could be removed. What do you think?

Ok, I hadn't realized that clicking on a button that has `tabIndex={-1}` would still focus it.  Yes I think you're right that this condition can be removed now because of the mousedown handler.

Reply:
Thanks. It is updated here 71713cfed122d7b4931f506e98892b7c521a733c.